### ISSUE TITLE
glob no longer returns sub-directories

### ISSUE THREAD

**Issue by brl0** [dask/gcsfs#287]
**What happened**:
Using version 0.7.1, `fs.glob` no longer returns sub-directories

**What you expected to happen**:
Using version 0.6.2, `fs.glob` returns sub-directories, which is the expected behavior.

**Minimal Complete Verifiable Example**:

```python
from fsspec.core import get_fs_token_paths
path = "gs://bucket"
fs, fs_token, paths = get_fs_token_paths(path, storage_options = {"token": "cloud"})
globbed = set(fs.glob(f"{path}/*"))
listed = set(fs.ls(path))
assert globbed == listed
```

**Anything else we need to know?**:
This does not seem to list sub-directories either:
`fs.find(path, withdirs=True)`

**Environment**:

- Dask version: N/A
- Python version: 3.7.8
- Operating System:debian:10.3-slim
- Install method (conda, pip, source): conda

**Comment by martindurant** [comment]
This is confirmed.

This is due to the change in `find` to make it more efficient for deeply nested directories (also in s3fs).

I believe the best thing to do would be to take the results of `find` to produce directory entries from the set of files, and also and back-fill the directory cache.
i.e. something like, 
```
dirs = set(self._parent(fn) for fn in out)
```

**Comment by martindurant** [comment]
Here is an implementation of that for s3: https://github.com/dask/s3fs/pull/371
I'll try to round that off, fix tests etc., and port here if successful. This might take a little time.

**Comment by martindurant** [comment]
Fixes by #289

**Comment by brl0** [comment]
@martindurant, thanks for the quick response and fix, I really appreciate it.
